automata/pda/pda.py

```diff
@@ -24,6 +40,280 @@ class PDA(Automaton, metaclass=abc.ABCMeta):
+                    fontsize=font_size_str,
+                    color=color.to_string(hex=True),
+                    penwidth="2.5",
+                )
+
+            graph = self._create_stack_diagram(
+                input_path, graph, start_color, end_color, font_size_str, arrow_size_str
+            )
+
+        edge_labels = defaultdict(list)
+        for from_state, to_state, symbol in self.iter_transitions():
+            if is_edge_drawn[from_state, to_state, symbol]:
+                continue
+
+            from_node = self._get_state_name(from_state)
+            to_node = self._get_state_name(to_state)
+            label = self._get_edge_name(*symbol)
+            edge_labels[from_node, to_node].append(label)
+
+        for (from_node, to_node), labels in edge_labels.items():
+            graph.add_edge(
+                from_node,
+                to_node,
+                label="\n".join(sorted(labels)),
+                arrowsize=arrow_size_str,
+                fontsize=font_size_str,
+            )
+
+        # Set layout
+        graph.layout(prog=layout_method)
+
+        # Write diagram to file. PNG, SVG, etc.
+        if path is not None:
+            save_path_final: pathlib.Path = pathlib.Path(path)
+
+            format = (
+                save_path_final.suffix.split(".")[1] if save_path_final.suffix else None
+            )
+
+            graph.draw(
+                path=save_path_final,
+                format=format,
+            )
+
+        return graph
+
+    def _create_stack_diagram(
+        self,
+        input_path: List[Tuple[PDAConfiguration, PDAConfiguration]],
+        graph: pgv.AGraph,
+        start_color: coloraide.Color,
+        end_color: coloraide.Color,
+        font_size: str,
+        arrow_size: str,
+    ) -> pgv.AGraph:
+        """
+        Constructs stack for all the transitions in the `input_path` and
+        adds the constructed stacks into `graph`. Returns the same `graph`
+        """
+        from_node = input_path[0][0]
+        label = " | ".join(reversed(from_node.stack))  # type: ignore[arg-type]
+        graph.add_node(
+            from_node,
+            label=f" | {label}",
+            shape="record",
+            arrowsize=arrow_size,
+            fontsize=font_size,
+        )
+
+        interpolation = coloraide.Color.interpolate(
+            [start_color, end_color], space="srgb"
+        )
+
+        for i, (c, n) in enumerate(input_path, start=1):
+            from_node = n
+            label = " | ".join(reversed(from_node.stack))  # type: ignore[arg-type]
+
+            color = interpolation(i / len(input_path))
+
+            graph.add_node(
+                from_node,
+                label=f" | {label}",
+                shape="record",
+                arrowsize=arrow_size,
+                fontsize=font_size,
+            )
+            graph.add_edge(
+                c,
+                n,
+                label=f"<<b>[<i>#{i}</i>]</b>>",
```

Comment:
In the edges, it doesn't look like the closing `]` is being displayed. Could something weird be going on here?

Reply:
The closing `]` is not displayed even in DFA.

![image](https://github.com/caleb531/automata/assets/55694469/6ea818cf-13f9-4e1d-b20c-a0f3c44bf271)


Not sure but maybe its a bug with `pygraphviz`?

Reply:
Oh hmmm that's a little strange. I'll play with the current version locally to see what's up

Reply:
Maybe this is a graphviz thing, but it looks fine on my machine:
![image](https://github.com/caleb531/automata/assets/1345068/b15d4d4d-dfca-4fa9-8315-3fdc0cc07d41)

Will go ahead and resolve.